nemo/collections/vlm/inference/base.py

```diff
@@ -0,0 +1,129 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+import pytorch_lightning as pl
+import torch
+import torch.distributed
+from megatron.core.inference.common_inference_params import CommonInferenceParams
+from megatron.core.inference.model_inference_wrappers.inference_wrapper_config import InferenceWrapperConfig
+from transformers import AutoProcessor
+
+import nemo.lightning as nl
+from nemo.collections import vlm
+from nemo.collections.vlm.inference.vlm_engine import VLMEngine
+from nemo.collections.vlm.inference.vlm_inference_controller import VLMTextGenerationController
+from nemo.collections.vlm.inference.vlm_inference_wrapper import VLMInferenceWrapper
+
+
+def _setup_trainer_and_restore_model(path: str, trainer: nl.Trainer, model: pl.LightningModule):
+    """Setup trainer and restore model from path"""
+    fabric = trainer.to_fabric()
+    model = fabric.load_model(path, model)
+    return model
+
+
+def setup_inference_wrapper(
+    model,
+    tokenizer,
+    params_dtype: torch.dtype = torch.bfloat16,
+    inference_batch_times_seqlen_threshold: int = 1000,
+):
+    """Set up inference wrapper for the model"""
+    config = model.config
+
+    mcore_model = model.module.cuda()
+    mcore_model = mcore_model.to(params_dtype)
+
+    inference_wrapped_model = VLMInferenceWrapper(
+        mcore_model,
+        InferenceWrapperConfig(
+            hidden_size=config.language_model_config.hidden_size,
+            params_dtype=params_dtype,
+            inference_batch_times_seqlen_threshold=inference_batch_times_seqlen_threshold,
+            padded_vocab_size=tokenizer.vocab_size,
+        ),
+    )
+
+    return inference_wrapped_model
+
+
+def setup_model_and_tokenizer(
+    path: str,
+    trainer: Optional[nl.Trainer] = None,
+    params_dtype: torch.dtype = torch.bfloat16,
+    inference_batch_times_seqlen_threshold: int = 1000,
+):
+    """Set up model and tokenizer"""
+    model_id = "meta-llama/Llama-3.2-11B-Vision-Instruct"
+
+    processor = AutoProcessor.from_pretrained(model_id)
+    tokenizer = processor.tokenizer
+    config = vlm.MLlamaConfig11BInstruct()
+    model = vlm.MLlamaModel(config, tokenizer=tokenizer)
+    _setup_trainer_and_restore_model(path=path, trainer=trainer, model=model)
+
+    inference_wrapped_model = setup_inference_wrapper(
+        model, tokenizer, params_dtype, inference_batch_times_seqlen_threshold
+    )
+
+    return inference_wrapped_model, processor
+
+
+def generate(
+    model: VLMInferenceWrapper,
```

Comment:
rename this model to wrapped model, just to be clear it needs to be wrapped first